third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp

```diff
@@ -817,6 +815,153 @@ SmallVector<unsigned, 2> getTilesPerWarp(Value aScale, Value bScale) {
   return {1, 1};
 }
 
+class DecomposeAMDScaledBlocked final : public ttg::DecomposeScaledBlocked {
+public:
+  DecomposeAMDScaledBlocked(MLIRContext *context, PatternBenefit benefit = 1)
+      : ttg::DecomposeScaledBlocked(context, benefit) {}
+  using TensorValue = TypedValue<RankedTensorType>;
+
+  LogicalResult matchAndRewrite(triton::DotScaledOp dotOp,
+                                PatternRewriter &rewriter) const override {
+    if (!dotOp.getLhsKPack() || !dotOp.getRhsKPack())
+      return failure();
+
+    RankedTensorType oldRetType = dotOp.getType();
+    if (!isa_and_nonnull<BlockedEncodingAttr>(oldRetType.getEncoding()))
+      return rewriter.notifyMatchFailure(
+          dotOp, "expected blocked encoding result tensor");
+    unsigned rank = oldRetType.getRank();
+    if (rank == 3)
+      return rewriter.notifyMatchFailure(dotOp, "NYI: 3d case");
+
+    TensorValue aScale = dotOp.getAScale();
+    TensorValue bScale = dotOp.getBScale();
+    if (aScale && bScale)
+      return rewriter.notifyMatchFailure(dotOp, "NYI: both LHS and RHS scale");
+
+    ScaleDotElemType aElemType = dotOp.getAElemType();
+    ScaleDotElemType bElemType = dotOp.getBElemType();
+    auto supportsTypes = [](ScaleDotElemType elemType) {
+      return elemType == ScaleDotElemType::E2M1 ||
+             elemType == ScaleDotElemType::E4M3 ||
+             elemType == ScaleDotElemType::E5M2 ||
+             elemType == ScaleDotElemType::BF16 ||
+             elemType == ScaleDotElemType::FP16;
+    };
+    if (!supportsTypes(aElemType) || !supportsTypes(bElemType))
+      return rewriter.notifyMatchFailure(dotOp, "NYI: mxfp6 operand");
+
+    bool useFp16 = aElemType == ScaleDotElemType::FP16 ||
+                   bElemType == ScaleDotElemType::FP16;
+    auto loc = dotOp.getLoc();
+
+    auto cvtDotOperand = [&](TensorValue v, int opIdx) -> TensorValue {
+      auto *ctx = rewriter.getContext();
+      auto retEnc = dotOp.getType().getEncoding();
+      auto vType = v.getType();
+      auto encoding = ttg::DotOperandEncodingAttr::get(ctx, opIdx, retEnc,
+                                                       vType.getElementType());
+      auto retTy = vType.cloneWithEncoding(encoding);
+      return rewriter.create<ttg::ConvertLayoutOp>(loc, retTy, v);
+    };
+
+    auto scaledA = scaleArg(rewriter, dotOp, 0, useFp16);
+    scaledA = cvtDotOperand(scaledA, 0);
+    auto scaledB = scaleArg(rewriter, dotOp, 1, useFp16);
+    scaledB = cvtDotOperand(scaledB, 1);
+
+    auto newDot =
+        rewriter.create<tt::DotOp>(loc, scaledA, scaledB, dotOp.getC());
+
+    rewriter.replaceOpWithNewOp<ttg::ConvertLayoutOp>(dotOp, oldRetType,
+                                                      newDot);
+    return success();
+  }
+
+  TensorValue scaleArg(PatternRewriter &rewriter, DotScaledOp dotOp, int opIdx,
```

Comment:
There are quite some duplication in this method from the base class. Can we define some internal functions to wrap smaller parts so we can customize only those?

Reply:
Hi @lezcano, could you take some time to review https://github.com/triton-lang/triton/pull/8078? It's an NFC to expose `DecomposeScaledBlocked` as @antiagainst mentioned. Btw, toggling on "Hide whitespace" might make the review easier, as there are many indent changes.

Reply:
Thanks @knwng! With two NFC changes landed and after the rebase, this looks better! I think we can do another pre-step: peel off the definition of the two new hardware ISA related scaled upcast ops and their associated llvm lowering into its separate pull request to further reduce the surface area of this pull request. Can you do that? Thanks!